src/game/WorldSession.cpp

```diff
@@ -102,17 +99,16 @@ WorldSession::~WorldSession()
         LogoutPlayer(true);
 
     /// - If have unclosed socket, close it
-    if (m_Socket)
-    {
-        m_Socket->CloseSocket();
-        m_Socket->RemoveReference();
-        m_Socket = nullptr;
-    }
+    if (!m_Socket->IsClosed())
+        m_Socket->Close();
 
     ///- empty incoming packet queue
-    WorldPacket* packet = nullptr;
-    while (_recvQueue.next(packet))
+    while (!m_packetQueue.empty())
```

Comment:
Hello.  Thank you for taking the time to review this.
> Are you sure nothing is being written to the queue at this point in time?

Actually you are right.  A new lock is needed there.  I added a new commit for it here: https://github.com/namreeb/mangos-wotlk/commit/f2b43bfcd47071f4a40024ba0e0074979e94723a

We are making the switch to boost, but I believe we want to use C++11 types whenever possible.  That is why I did not use their container.